### DISCUSSION TITLE
How to transpose a tensor and have the transpose reflected in contiguous memory with CUDA backend

### DISCUSSION THREAD

**Discussion by balisujohn** [ggml-org/ggml#684] [Q&A] [answered]
I have a contiguous tensor of shape `[18,64,16,4]` If I print out the first and last 3 memory locations I get
```
0.00137997
0.00800999
0.0252284
0.18625
-0.075262
-0.00594251
```
I want to transpose this tensor and have the transpose reflected in contiguous memory. So I try the code:
```
KQV = ggml_transpose(ctx0, KQV);
KQV = ggml_cont_3d(ctx0,ggml_view_1d(ctx0, KQV, 64*18*16*4,0),64,18,64);
```
and I get the same results when printing out the first and last 3 values from memory:
```
0.00137997
0.00800999
0.0252284
0.18625
-0.075262
-0.00594251
```
How can I get a transpose that's actually reflected in contiguous memory for the first 2 dimensions of a 4d tensor? I should add that this is with a cuda backend, and `ggml_cont_4d()` doesn't work with 4d tensors for the cuda backend.

**Comment by balisujohn** [comment, marked as answer]
This strategy seems to work actually:
```
            KQV = ggml_reshape_3d(ctx0, KQV, 18,64,64);
            KQV = ggml_permute(ctx0, KQV, 1,0,2,3);
            KQV = ggml_cont_3d(ctx0, KQV, 64,18,64);
            KQV = ggml_reshape_4d(ctx0, KQV, 64,18,16,4);
```

**Reply by balisujohn** [reply]
~~(though im not getting exactly the transpose I want still even when I tried 1,0,2,3, but that's a separate issue)~~
updated the version above so it now works correctly!

**Reply by ggerganov** [reply]
```c
KQV = ggml_cont(ggml_transpose(ctx0, KQV));
```

should also work
